packages/e2e-tests/src/Yarn.ts

```diff
@@ -0,0 +1,131 @@
+import * as os from 'os';
+import * as path from 'path';
+import { Cache } from './Cache';
+import * as core from '@actions/core';
+import * as glob from '@actions/glob';
+import * as artifact from '@actions/artifact';
+import { SpawnSync } from './SpawnSync';
+import { Repository } from './Repository';
+import { EnvVars } from './Action';
+
+class Yarn {
+	constructor(
+		private readonly repo: Repository,
+		private readonly spawnSync: SpawnSync,
+		private readonly cache: Cache
+	) {}
+
+	public async install({ frozenLockfile }: { frozenLockfile: boolean }): Promise<Yarn> {
+		const args = ['install'];
+
+		if (frozenLockfile) {
+			args.push('--frozen-lockfile');
+		}
+
+		await this.call(args, ['node_modules', '*/node_modules']);
+
+		return this;
+	}
+
+	public async build(): Promise<Yarn> {
+		await this.call(['build'], ['build']);
+		return this;
+	}
+
+	public async test(envVars: EnvVars): Promise<Yarn> {
+		// TODO: once e2e-tests package is extracted, update this
+
+		const caRoot = this.spawnSync.call('mkcert', ['-CAROOT'], { stdout: 'pipe' }).stdout.trim();
+
+		const reportPath = path.resolve(os.tmpdir(), 'playwright-report');
+
+		const env = {
+			NODE_EXTRA_CA_CERTS: `${caRoot}/rootCA.pem`,
+			PLAYWRIGHT_HTML_REPORT: reportPath,
+			...envVars,
+		} as const;
+
+		// if docker cache will become available, restore should be called here
+
+		const buffer = this.spawnSync.call('yarn', ['workspace', '@eventespresso/e2e', 'playwright', 'test'], {
+			env,
+		});
+
+		// if docker cache will become available, save should be called here
+
+		if (buffer.status !== 0) {
+			await this.saveTestReport(reportPath);
+		}
+
+		return this;
+	}
+
+	private async saveTestReport(reportPath: string): Promise<void> {
+		core.notice(`Saving Playwright report found at: ${reportPath}`);
+
+		const client = artifact.create();
+
+		const pattern = path.resolve(reportPath, '**/*');
+
+		const globber = await glob.create(pattern);
+
+		const files = await globber.glob();
+
+		const reportName = `playwright-report-run-${process.env.GITHUB_RUN_NUMBER}-attempt-${process.env.GITHUB_RUN_ATTEMPT}`;
+
+		let response = undefined;
+
+		try {
+			response = await client.uploadArtifact(reportName, files, reportPath, {
+				continueOnError: true,
+				retentionDays: 7,
+			});
+		} catch (error) {
+			core.error(error as string);
+			return;
+		}
+
+		if (response.failedItems.length > 0) {
+			core.error(`Failed to upload some artifact items: \n${response.failedItems.join(', ')}`);
```

Comment:
should this include the `reportName`?

Reply:
We could, but imho, it won't be very beneficial. I don't think Playwright would be able to render a report with missing files. This error log here is mainly to make it easier to debug if report cannot be saved as artifact propertly.